SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,436 @@
+    case let atp as AddressToPointerInst:
+      if let result = result, atp.operand != result {
+        self.result = nil
+        return .abortWalk
+      }
+      // TODO: instead of checking type equality perhaps check
+      //       layout compatibility, if possible?
+      if addrType == atp.operand.type, path.isEmpty {
+        self.result = atp.operand
+        return .continueWalk
+      }
+    case is BlockArgument, is MarkDependenceInst, is CopyValueInst,
+         is StructExtractInst, is TupleExtractInst, is StructInst, is TupleInst:
+      return walkUpDefault(value: value, path: path)
+    default:
+      break
+    }
+
+    self.result = nil
+    return .abortWalk
+  }
+
+  var walkUpCache = WalkerCache<Path>()
+}
+
+struct PointerIdentificationWalker {
+  private var walker = PointerIdentificationUseDefWalkerInternal()
+
+  mutating func compute(_ atp: PointerToAddressInst) -> Value? {
+    walker.addrType = atp.type
+    walker.result = nil
+    _ = walker.walkUp(value: atp.operand, path: SmallProjectionPath())
+    return walker.result
+  }
+}
+
+fileprivate struct AccessPathWalkerInternal : AddressUseDefWalker {  
+  var result: AccessPath? = nil
+  private var _beginAccess: BeginAccessInst? = nil
+  var scope: EnclosingScope? {
+    if let ba = _beginAccess {
+      return .scope(ba)
+    } else {
+      guard let accessPath = result else { return nil }
+      return .base(accessPath.base)
+    }
+  }
+
+  private var pointerWalker = PointerIdentificationWalker()
+  
+  mutating func start() {
+    result = nil
+    _beginAccess = nil
+  }
+
+  struct Path : SimpleWalkingPath {
+    let projectionPath: SmallProjectionPath
+
+    // Tracks whether an `index_addr` instruction was crossed.
+    // It should be (FIXME: check if it's enforced) that operands
+    // of `index_addr` must be `tail_addr` or other `index_addr` results.
+    let indexAddr: Bool
+
+    func with(projectionPath: SmallProjectionPath) -> Self {
+      return Self(projectionPath: projectionPath, indexAddr: indexAddr)
+    }
+
+    func with(indexAddr: Bool) -> Self {
+      return Self(projectionPath: projectionPath, indexAddr: indexAddr)
+    }
+
+    func merge(with other: Self) -> Self {
+      return Self(
+        projectionPath: projectionPath.merge(with: other.projectionPath),
+        indexAddr: indexAddr || other.indexAddr
+      )
+    }
+  }
+  
+  mutating func rootDef(address: Value, path: Path) -> WalkResult {
+    
+    func answer(_ kind: AccessBaseKind?) -> WalkResult {
+      if let kind = kind {
+        self.result =  AccessPath(
+          base: AccessBase(baseAddress: address, baseKind: kind)!,
+          projectionPath: path.projectionPath
+        )
+        return .continueWalk
+      } else {
+        self.result = nil
+        return .abortWalk
+      }
+    }
+
+    // Try identifying the address a pointer originates from
+    if let p2ai = address as? PointerToAddressInst,
+       let newAddress = pointerWalker.compute(p2ai) {
```

Comment:
@eeckstein @Angelogeb 
The pointer walker utility should be as separate from the address walker as we can get. But the reason the pointer walker needs to be called from the address walker is to handle global addressors. If the access is a global variable, then the address walker needs to recognize that for correctness. So, the address walker at least needs to attempt to find a single common address for any pointer-to-address operation before returning an Unidentified result.

The SIL verifier should verify that this attempt always succeeds for global variable access. Ideally, we should make global variable access more explicit in SIL so we're not relying on recognizable patterns from SILGen.

This is the bad SIL pattern, where the address walker needs to a pointer walk:

```
%globalPtr = apply %globalAddressor
%globalAdr = pointer_to_address
%access = begin_access %globalAdr
load %access
```

It would be great to disallow that SIL.

Whereas, this is a normal pattern where the address walker can safely stop at pointer_to_address and report an Unidentified base. But the results still are more useful here if it does the pointer walk and finds the global base:

```
%globalAdr = global_adr
%access = begin_access %globalAdr
%globalPtr = address_to_pointer
%adr = pointer_to_address
load %adr
```

Reply:
@atrick I agree. This is an unfortunate special case we need to handle. But I'm wondering if we could do it in a simpler way. Is it ever the case that this pattern is more complex than in your example (especially in the mandatory pipeline, where we rely on the pattern to be recognized)?
If we can assume that it's always a sequence of `apply` -> `pointer_to_address` then we can just handle this case and don't need a walker.

If we make AccessBase a protocol (but still keeping the AccessBaseKindEnum), then the conforming `var PointerToAddress.kind` can deal with this and return `.Global` in case its operand is an apply of a global addressor function or `.Pointer` otherwise.

Reply:
The current implementation of pointer identification only identifies a _common_ address, possibly crossing phi "webs". This handles the `address_to_pointer` -> `pointer_to_address` sequence transparently, but also [this pathological case](https://github.com/apple/swift/blob/8f9c5339542b17af9033f51ad7a0b95a043cad1b/test/SILOptimizer/access_storage_analysis_ossa.sil#L669-L705), and nothing more.

I think making `AccessBase` a protocol is not ideal since some of them (such as "argument" and "yield") aren't bases in general.